haddock-api/src/Haddock/Interface/LexParseRn.hs

```diff
@@ -84,21 +84,27 @@ processModuleHeader dflags pkgName gre safety mayStr = do
 -- fallbacks in case we can't locate the identifiers.
 --
 -- See the comments in the source for implementation commentary.
-rename :: DynFlags -> GlobalRdrEnv -> Doc RdrName -> ErrMsgM (Doc Name)
+rename :: DynFlags -> GlobalRdrEnv -> Doc NsRdrName -> ErrMsgM (Doc Name)
 rename dflags gre = rn
   where
     rn d = case d of
       DocAppend a b -> DocAppend <$> rn a <*> rn b
       DocParagraph doc -> DocParagraph <$> rn doc
-      DocIdentifier x -> do
+      DocIdentifier (NsRdrName ns x) -> do
         -- Generate the choices for the possible kind of thing this
         -- is.
         let choices = dataTcOccs x
         -- Try to look up all the names in the GlobalRdrEnv that match
         -- the names.
         let names = concatMap (\c -> map gre_name (lookupGRE_RdrName c gre)) choices
 
-        case names of
+        -- Filter according to the namespace
+        let filterNamespace = case ns of
+                                Value -> filter (not . isTyConName)
+                                Type -> filter isTyConName
+                                None -> id
+
+        case filterNamespace names of
```

Comment:
Instead of looking up all the possibilities and filtering, let's look up only what we're interested in. Take a look at the implementation of `dataTcOccs` for that.